File: samples/nodejs/eftl/subscriber.js
/*
 * Copyright (c) 2013-$Date: 2020-09-24 12:20:18 -0700 (Thu, 24 Sep 2020) $ TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

'use strict'

const eFTL = require('eftl');
const fs = require('fs');

const url = (process.argv[2] || "ws://localhost:8585/channel");

console.log(eFTL.getVersion());

// Set the server certificate chain for secure connections.
// Self-signed server certificates are not supported.
//
//eFTL.setTrustCertificates(fs.readFileSync('./ca.pem'));

// Start a connection to the eFTL server.
//
// The username and password are required when connecting
// to an eFTL server that has been configured for JAAS.
//
// The clientId uniquely defines the eFTL client. If not
// provided the eFTL server will generate one for the client.
//
// The onConnect function is invoked following a successful
// connection to the eFTL server.
//
// The onDisconnect function is invoked following a failed
// connection attempt to the eFTL server.
//
eFTL.connect(url, {
    username: 'user',
    password: 'password',
    trustAll: false,
    clientId: 'sample-node-client',
    onConnect: connection => {
        console.log('Connected to eFTL server');
        subscribe(connection);
    },
    onDisconnect: (connection, code, reason) => {
        console.log('Disconnected from eFTL server: ' + reason);
    },
    onError: (connection, code, reason) => {
        console.log('Error from eFTL server: ' + reason);
    }
});

// Subscribe to eFTL messages using a durable subscription.
//
// This subscription defines a matcher that will match published 
// messages containing a field named "type" with a value of "hello".
//
// The durable name is used to create a durable subscription. When
// setting a durable name also set the clientId in the connect
// method as durable subscriptions are mapped to specific clientIds.
//
// The subscription's message acknowledgment mode can be set to
// 'auto', 'client', or 'none'. When set to 'client' the consumed
// messages must be explicitly acknowledged. The default message
// acknowledgment mode is 'auto'.
//
// The onSubscribe function is invoked following a successful
// subscription.
//
// The onError function is invoked following a failed subscription.
//
// the onMessage function is invoked whenever a published message
// is received by the subscription.
//
function subscribe(connection) {
    connection.subscribe({
        matcher: `{"type":"hello"}`,
        durable: 'sample-durable',
        ack: 'client',
        onSubscribe: id => {
            console.log('Subscribed');

            // Unsubscribe and disconnect after 30 seconds.
            setTimeout(function() {
                // This will permanently remove a durable subscription.
                connection.unsubscribe(id);
                connection.disconnect();
            }, 30000);
        },
        onError: (id, code, reason) => {
            console.log('Subscription failed: ' + reason);
        },
        onMessage: message => {
            console.log('Received message: ' + message.toString());
        }
    });
}

File: eftl-golang-sdk/examples/subscriber/subscriber.go
//
// Copyright (c) 2001-$Date: 2017-03-03 16:29:04 -0800 (Fri, 03 Mar 2017) $ TIBCO Software Inc.
// All Rights Reserved. Confidential & Proprietary.
// For more information, please contact:
// TIBCO Software Inc., Palo Alto, California, USA
//

// This is an example of a basic eFTL client subscriber.

package main

import (
	"log"
	"os"

	"tibco.com/eftl"
)

var (
	url = "ws://localhost:8585/channel"
)

const (
	username  = ""
	password  = ""
	clientID  = "sample-go-client"
	durable   = "sample-durable"
	clientAck = false
	count     = 10
)

func main() {

	if len(os.Args) > 1 {
		url = os.Args[1]
	}
        log.Printf("%s : TIBCO eFTL Version: %s\n", os.Args[0], eftl.Version)

	// set the log level log file
        eftl.SetLogLevel(eftl.LogLevelDebug)
        eftl.SetLogFile("subscriber.log")

	// channel for receiving connection errors
	errChan := make(chan error, 1)

	// set connection options
	opts := eftl.DefaultOptions()
	opts.ClientID = clientID
	opts.Username = username
	opts.Password = password

	// connect to the server
	conn, err := eftl.Connect(url, opts, errChan)
	if err != nil {
		log.Printf("connect failed: %s", err)
		return
	}

	// close the connection when done
	defer conn.Disconnect()

	// channel for receiving subscription response
	subChan := make(chan *eftl.Subscription, 1)

	// channel for receiving published messages
	msgChan := make(chan eftl.Message, 1000)

	// create the message content matcher
	matcher := "{\"type\":\"hello\"}"

	// create the subscription options
	subopts := eftl.SubscriptionOptions{}

	if clientAck {
		subopts.AcknowledgeMode = eftl.AcknowledgeModeClient
	}

	// create the subscription
	//
	// always set the client identifier when creating a durable subscription
	// as durable subscriptions are mapped to the client identifier
	//
	err = conn.SubscribeWithOptionsAsync(matcher, durable, subopts, msgChan, subChan)
	if err != nil {
		log.Printf("subscribe failed: %s", err)
		return
	}

	var total int

	for {
		select {
		case sub := <-subChan:
			if sub.Error != nil {
				log.Printf("subscribe operation failed: %s", sub.Error)
				return
			}
			log.Printf("subscribed with matcher: %s", sub.Matcher)
			// unsubscribe, this will permanently remove a durable subscription
			defer conn.Unsubscribe(sub)
		case msg := <-msgChan:
			total++
			log.Printf("received message: %s", msg)

			if clientAck {
				conn.Acknowledge(msg)
			}

			if total >= count {
				return
			}
		case err := <-errChan:
			log.Printf("connection error: %s", err)
			return
		}
	}
}
